### PR TITLE
VIMC-3450: improved montagu-db restore

### PR DESCRIPTION
**Pull request by richfitz**
This PR documents the backup process a bit, and also updates the restore so that we run a postgres over the restore volume after dumping it out, which will speed up deployment.

This is _really_ hard to test, and it might take a couple of shots to get it right

### CODE REVIEW

**Comment by richfitz** [comment]
See also: https://github.com/vimc/montagu-bb8/pull/9

File: design.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+2. the **WAL (Write Ahead Log)** which can be replayed from a base backup to the present to restore the current database state
+
+The overall process is described in diagram form [here](https://github.com/vimc/montagu-machine/blob/master/docs/diagrams/Database%20backup.png)
+
+### The restore path
+
+The basic approach (without any real reference to the wrinkles that will appear below) is:
+
+1. On annex, `barman` dumps out a recovery directory (this is a base backup plus the WAL) using `barman recover` into the `montagu_db_volume` volume - this is our "nightly backup" and is controlled by cron so is error prone, and may stop running without warning.
+2. On annex, `bb8` ships this into the starport using `bb8 backup`, updating the directory on disk `~/starport/barman_to_starport` (this is the only backup target _from_ annex, even though everything else gets sent there).
```
Comment (EmmaLRussell) [inline design.md]: Add migration step - migrating to restore volume on annex from which Science is restored (and Production could be)..

File: barman-montagu [hunk 2]
```diff
@@ -82,8 +82,30 @@ def recover(id, wipe_first):
...
-    exec_safely(barman, ["update-nightly", DATABASE_NAME],
+    print("Dumping backup into nightly volume")
+    exec_safely(barman, ["barman", "recover", DATABASE_NAME, "latest",
+                         "/nightly/.incoming"],
                 stream=True, check=True)
+    migrate_nightly()
+    exec_safely(barman, ["update-nightly"], stream=True, check=True)
+
+
+def migrate_nightly():
```
Comment (EmmaLRussell) [inline barman-montagu]: 'Migrate' being an overloaded term, more clunky 'replay WAL' or similar might be more explicit

File: design.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
+## Montagu DB backup - design
```
[the remark this answers is not in the document]
Comment (richfitz) [reply design.md]: I've added a reference to `barman-montagu status` below in the base backups section

File: design.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+```
+barman-montagu barman backup montagu
+```
+
+This has implications for cleanup though, as our retention policy guarantees only the last 3 backups will be saved.
+
+### Off-site backup
+
+Our backup system copies the starport volume into AWS.  There is also a second copy of barman running there.
```
[the remark this answers is not in the document]
Comment (richfitz) [reply design.md]: This is linked now

File: design.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+3. On annex, `bb8` ships this into the starport using `bb8 backup`, updating the directory on disk `~/starport/barman_to_starport` (this is the only backup target _from_ annex, even though everything else gets sent there).
+4. On science (or another machine being restored) `bb8 restore` rsync's the contents of the starport barman into the `montagu_db_volume` container, then starts montagu.
+
+A base backup will take ~2hrs (as of 2020-02-20) and currently requires ~220GB.  If there are significant quantities of WAL logs then it will take longer.  Replaying the WAL can take up to an hour.
+
+Steps 1-3 are done using a cron script that lives in the bb8 repo `schedule-barman-montagu-nightly`.  This cron job can and has stopped without warning in the past - most recently when the Python installation disappeared during an Ubuntu upgrade (but see [Logging](#Logging) below).
+
+bb8 takes control of the actual process of preparing the recovery volume.  It writes out a cron nightly script that runs the barman `update-nightly` script and then `bb8 backup`
```
Comment (hillalex) [inline design.md]:
```
```

File: design.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+2. On annex, The nightly backup is then run through a copy of Postgres, using `vimc/montagu-db:latest`, so that the database is fully recovered (`barman revover` dumps the data that Postgres *can* restore, but it takes time for the WAL to be replayed over the base backup) - we're calling this step "replay wal"
+3. On annex, `bb8` ships this into the starport using `bb8 backup`, updating the directory on disk `~/starport/barman_to_starport` (this is the only backup target _from_ annex, even though everything else gets sent there).
+4. On science (or another machine being restored) `bb8 restore` rsync's the contents of the starport barman into the `montagu_db_volume` container, then starts montagu.
+
+A base backup will take ~2hrs (as of 2020-02-20) and currently requires ~220GB.  If there are significant quantities of WAL logs then it will take longer.  Replaying the WAL can take up to an hour.
+
+Steps 1-3 are done using a cron script that lives in the bb8 repo `schedule-barman-montagu-nightly`.  This cron job can and has stopped without warning in the past - most recently when the Python installation disappeared during an Ubuntu upgrade (but see [Logging](#Logging) below).
```
Comment (hillalex) [inline design.md]:
```suggestion
Steps 1-3 are done using a cron script that lives in the montagu-bb8 repo `schedule-barman-montagu-nightly`.  This cron job can and has stopped without warning in the past - most recently when the Python installation disappeared during an Ubuntu upgrade (but see [Logging](#Logging) below).
```

File: bin/update-nightly [hunk 7]
```diff
@@ -28,7 +28,11 @@ INCOMING_PATH=.incoming
...
 cd $NIGHTLY_PATH
 
-barman recover $DATABASE latest $INCOMING_PATH
+if [ ! -d "$NIGHTLY_PATH/$INCOMING_PATH" ]; then
+    echo "Did not find new restore in $NIGHTLY_PATH/$INCOMING_PATH"
+    exit 1
+fi
+
 rm -f $NIGHTLY_PATH/.barman-last-restore
 # This is the extglob syntax - read as "all paths except $INCOMING_PATH"
```
Comment (hillalex) [inline bin/update-nightly]:
```suggestion
barman recover $DATABASE latest $INCOMING_PATH
# This is the extglob syntax - read as "all paths except $INCOMING_PATH"
```

**Pull request merged**